renderer/backend/UniversalGpuBuffer.cpp

```diff
@@ -0,0 +1,158 @@
+/*****************************************************************************
+                    The Dark Mod GPL Source Code
+
+ This file is part of the The Dark Mod Source Code, originally based
+ on the Doom 3 GPL Source Code as published in 2011.
+
+ The Dark Mod Source Code is free software: you can redistribute it
+ and/or modify it under the terms of the GNU General Public License as
+ published by the Free Software Foundation, either version 3 of the License,
+ or (at your option) any later version. For details, see LICENSE.TXT.
+
+ Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+#include "precompiled.h"
+#include "UniversalGpuBuffer.h"
+#include "../tr_local.h"
+
+extern idCVarBool r_usePersistentMapping;
+
+void UniversalGpuBuffer::Init( GLenum type, GLuint size, GLuint alignment ) {
+	if( mMapBase ) {
+		Destroy();
+	}
+
+	mSize = ALIGN( size, alignment );
+	mAlign = alignment;
+	mType = type;
+
+	qglGenBuffers( 1, &mBufferObject );
+	qglBindBuffer( type, mBufferObject );
+
+	mUsePersistentMapping = r_usePersistentMapping && glConfig.bufferStorageAvailable;
+
+	if (mUsePersistentMapping) {
+		qglBufferStorage( type, mSize, nullptr, GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
+		mMapBase = ( byte* )qglMapBufferRange( type, 0, mSize, GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
+	} else {
+		qglBufferData( type, mSize, nullptr, GL_DYNAMIC_DRAW );
+		mMapBase = ( byte* )Mem_Alloc16( mSize );
+	}
+	mCurrentOffset = 0;
+	mLastLocked = 0;
+}
+
+void UniversalGpuBuffer::Destroy() {
+	if( !mBufferObject ) {
+		return;
+	}
+
+	for( auto it : mRangeLocks ) {
+		qglDeleteSync( it.fenceSync );
+	}
+	mRangeLocks.clear();
+
+	if (mUsePersistentMapping) {
+		qglBindBuffer( mType, mBufferObject );
+		qglUnmapBuffer( mType );
+	} else {
+		Mem_Free16( mMapBase );
+	}
+	qglDeleteBuffers( 1, &mBufferObject );
+	mBufferObject = 0;
+	mMapBase = nullptr;
+}
+
+void UniversalGpuBuffer::Lock() {
+	if( mCurrentOffset != mLastLocked ) {
+		LockRange( mLastLocked, mCurrentOffset - mLastLocked );
+		mLastLocked = mCurrentOffset;
+	}
+}
+
+byte * UniversalGpuBuffer::Reserve( GLuint size, bool precommit ) {
+	GLuint requestedSize = ALIGN( size, mAlign );
+	if( requestedSize > mSize ) {
+		common->Error( "Requested shader param size exceeds buffer size" );
+	}
+
+	if( mCurrentOffset + requestedSize > mSize ) {
+		if( mCurrentOffset != mLastLocked ) {
+			LockRange( mLastLocked, mCurrentOffset - mLastLocked );
```

Comment:
I'll add the requested comments to the class, which will hopefully make it clearer. But I do think the behaviour as it is is correct, as long as you keep to the contract of the interface - which unfortunately is a little awkward, but I haven't been able to find anything better.

The contract is basically:

1. You request a certain amount of storage from the buffer. It cannot immediately be marked as used for two reasons, one being the fallback from persistent buffers and the other being the backend design where it's virtually impossible to know how many draw calls you'll have in a batch until you actually start assembling them.
2. Once you've filled the requested memory with your data, and are ready to use it, you commit the data, which will mark it as used and, in case of fallback, actually upload it (not needed with persistent mapping).
3. Now you should use the data in draw calls before requesting any new storage.
4. Finally, this is where you could lock the used range. However, since fence syncs are expensive, we can't lock after every small draw batch, so by default a Lock is only issued once at the end of a frame. The only exception is if the buffer needs to wrap around.

If you stick with this contract, both of your outlined situations cannot happen - the locking will specifically ensure that overflow is not a problem, unless you make a single request that is larger than the entire buffer. If you don't stick with it, you can get into even more problems, I think. Unfortunately, I don't see a good way to enforce this contract or provide a better interface. I'm open to ideas. I do agree it needs documentation.

Reply:
The fences are use to ensure that when you wraparound, you wait for GL commands using the old data to complete before writing the old data. For it to happen, the fence must be issued after the **user has finished all rendering with the data**. Issuing fence without explicit `Lock` creates race condition, because the user does not know about it, and he issues draw commands after the fence.

Reply:
Speaking of the *overflow* situation, it seems that it is not supported. I.e. if it happens, geometry will disappear randomly or just render incorrectly. I'm afraid in order to support overflow, the user must know when his not-yet-locked allocations has filled the whole buffer, so that he could issue draw commands before pushing more. There is no interface for it, and it is not easy to do, since requires assistance from the user.

Note that I'm not speaking about the case when single allocation is larger than buffer --- that's clearly forbidden and even checked for.

Reply:
The contract is that you must have committed *and rendered* any previous data before you request new data from the buffer, that's the contract. Therefore, the issuing of the fence is correct because the as-yet unlocked data *has been used in a draw call* per the contract. And that's also why the overflow works - you land back at the beginning of a buffer and then explicitly wait for any fences that overlap with the requested range.

Obviously, you can break that contract, like request some data, fill it, then commit it and immediately request another block and only issue draw calls after both. And this is a scenario that I wish I could support, because there are possibly instances where it could be useful to have prepared multiple blocks for a single draw call. But it is simply forbidden due to the very reason you outlined.

Again, if you have ideas for a better interface, I'm open to it. I've thought about several alternatives, and in the end, they've all been bad for various reasons.

Reply:
One rough idea I had is I could perhaps fuse the DrawBatchExecutor and the ShaderParamsBuffer together. Right now, they are independent abstractions on top of the UniversalGpuBuffer, but you'd basically always use them in tandem. And if that's reflected in their combined interface, it would be much harder to misuse them. You could still misuse the UniversalGpuBuffer, but that's a lower primitive that you'd not normally use directly.

Reply:
I think it's better to have usage patterns of this class crystal clear. It is usually easier to understand something smaller, than the whole mix of three classes. And you said that you think about migrating `VertexCache` to this buffer, so it is important to make it clear.

Ok, I see it now: user can only Reserve+Commit+Draw+Reserve+Commit+Draw+... in this strict sequence, and Lock is optional and used only for better performance. Basically, only one allocation can be alive at any moment. Sounds like a consumer of immediate-style data, no reuse between allocations is allowed.
How can VertexCache be implemented then? Would you sort all surfaces by vertex pointer to ensure that all draws of the same mesh go at the same time, or load the mesh many times?

As for suggestions, I made one in the original post. You can allow the situation when `mLastLocked > mCurrentOffset`, making it fully circular. If it happens, then you have to issue two fences in `Lock` instead of one, but that's not a big problem. If you do so, then the user can reserve several allocations in a row, then commit them in arbitrary order, then draw various stuff using them in arbitrary order, and finally `Lock`. Of course, 1) locking would become mandatory, and 2) total size of unlocked allocations must not exceed buffer size.

Regarding the last condition, I think it can be lifted away if you return NULL on allocation with overflow. If the user issues all draws before allocating more stuff (as you expect it now), then it is enough for him to check for NULL, and call `Lock`  + retry `Reserve` if it happened. But for something like VertexCache, caller can resize the whole buffer for instance.

I'm not sure if it is better. I guess this question would not appear if the class had a comment from the very beginning.

Reply:
After this discussion, I'm beginning to favour the approach of abandoning a circular buffer and using the 'triple-range' concept loosely outlined in the other thread. Combined with an overhaul of the DrawBatchExecutor/ShaderParamsBuffer, it could potentially result in a cleaner interface with fewer chances of misuse. And you're right, the circular pattern right now would have some issues to be used in the VertexCache...

But it's not a trivial change, so I'll have to think about it and experiment a little.

Reply:
I have committed a reworked `GpuBuffer` and `DrawBatchExecutor`, the latter also taking over and replacing the responsibilities of `ShaderParamsBuffer`. I also added some comments about its workings.

As proposed, the new `GpuBuffer` is no longer a circular buffer, but rather divided into three frame regions (and an optional static data region), where each frame writes to particular region, and then at the end of the frame that region is locked as a whole and you switch to the next region. It should make the whole thing conceptually simpler, as there's no wraparound to worry about.

The `DrawBatchExecutor` will now give you a simple struct for a DrawBatch with two arrays that you have to fill with the shader params and the drawSurfs you want to draw and will handle the `GpuBuffer` for the UBO automatically in the background for you. It still doesn't have the greatest interface, but I think hopefully it is a bit clearer and much harder to misuse. Let me know what you think.